### PR TITLE
Make export_data task work with remote storage (bug 1191878)

### PR DESCRIPTION
**Pull request by diox**
https://bugzilla.mozilla.org/show_bug.cgi?id=1191878

### CODE REVIEW

**Comment by diox** [comment]
WIP: it's not entirely working right now, and tests don't show that, so I need to fix it.

File: mkt/webapps/tasks.py [hunk 1]
```diff
@@ -383,45 +382,85 @@ def export_data(name=None):
...
 
-    # Put some .txt files in place.
-    cmd = ['tar', 'czf', target_file, '-C',
-           settings.DUMPED_APPS_PATH] + files
-    task_log.info(u'Creating dump {0}'.format(target_file))
-    subprocess.call(cmd)
-    return target_file
+    # Clean-up.
+    local_target_file.close()
+    # rm_directory(local_source_dir)
```
Comment (diox) [inline mkt/webapps/tasks.py]: I need to uncomment that

File: mkt/site/storage_utils.py [hunk 2]
```diff
@@ -150,12 +150,17 @@ def walk_storage(path, topdown=True, onerror=None, followlinks=False,
...
+    Copy a path (src_path) from a storage (src_storage) to a path (dst_path)
+    on a different storage (dst_storage).
+
+    Defaults to copying from the local storage to the private storage.
     """
-    with open(src_path) as src_f, storage.open(dest_path, 'w') as dest_f:
-        shutil.copyfileobj(src_f, dest_f)
+    with src_storage.open(src_path) as src_f, \
+            dst_storage.open(dst_path, 'w') as dst_f:
+        shutil.copyfileobj(src_f, dst_f)
```
Comment (robhudson) [inline mkt/site/storage_utils.py]: I feel like I've written this same code before. But maybe it was another method and not this one.

File: mkt/webapps/tasks.py [hunk 3]
```diff
@@ -333,8 +336,6 @@ def unindex_webapps(ids, **kw):
 @task
 def dump_app(id, **kw):
     from mkt.webapps.serializers import AppSerializer
-    # Because @robhudson told me to.
```
Comment (robhudson) [inline mkt/webapps/tasks.py]: The blame is finally lifted.

File: mkt/webapps/tasks.py [hunk 4]
```diff
@@ -383,45 +382,85 @@ def export_data(name=None):
...
         name = today
-    root = settings.DUMPED_APPS_PATH
-    directories = ['apps']
-    for directory in directories:
-        rm_directory(os.path.join(root, directory))
-    files = directories + compile_extra_files(date=today)
+
+    # Clean up the path where we'll store the individual json files from each
+    # app dump.
+    for dirpath, dirnams, filenames in walk_storage(settings.DUMPED_APPS_PATH):
```
Comment (robhudson) [inline mkt/webapps/tasks.py]: How does `walk_storage` know which bucket to walk?

Comment (diox) [inline mkt/webapps/tasks.py]: It defaults to `private_storage`. I debated explicitly passing it as argument anyway, but thought this function was hard enough to read as it is.

File: mkt/webapps/tasks.py [hunk 5]
```diff
@@ -383,45 +382,85 @@ def export_data(name=None):
...
+    task_log.info(u'Creating dump {0}'.format(local_target_file.name))
+    subprocess.call(cmd)
 
-    if not os.path.exists(target_dir):
-        os.makedirs(target_dir)
+    # Now copy the local tgz to the public storage.
+    remote_target_filename = os.path.join(
+        settings.DUMPED_APPS_PATH, 'tarballs', '%s.tgz' % tarball_name)
+    copy_to_storage(local_target_file.name, remote_target_filename,
+                    dst_storage=public_storage)
```
Comment (robhudson) [inline mkt/webapps/tasks.py]: you don't need `src_storage` here?

File: mkt/site/storage_utils.py (hunk 2 above)
Comment (robhudson) [inline mkt/site/storage_utils.py]: src_storage defaulting to local storage seems odd to me when almost everythign else defaults to private.

**Comment by robhudson** [comment]
r+wc. I didn't think much on all the paths strings but I assume they're good.

File: mkt/webapps/tasks.py (hunk 5 above)
Comment (diox) [inline mkt/webapps/tasks.py]: `src_storage` defaults to `local_storage` so like above I didn't set arguments I didn't need to set

File: mkt/site/storage_utils.py (hunk 2 above)
Comment (diox) [inline mkt/site/storage_utils.py]: It seems strange, but it actually makes sense as the method was originally designed to copy stuff from local filesystem to the remote storage. The remote storage does default to `private_storage`.

**Comment by diox** [comment]
Tests are now passing when configured with a remote storage, and I've added the explicit `storage=private_storage` argument to `walk_storage()` calls.

**Pull request merged**